### PR TITLE
ArcInvestigation level assays

### PR DESCRIPTION
**Pull request by Freymaurer**
- Add parent option for ArcAssay and ArcStudy
- Move full ArcAssays out of ArcStudy to ArcInvestigation

### CODE REVIEW

**Comment by HLWeil** [comment]
CI bugged?

image

Everything works for me locally. I even tried cloning the repo again (which also worked).

**Comment by HLWeil** [comment]
Also: The `erroneous line` does not even exist: 

image


https://github.com/nfdi4plants/ARCtrl/blob/developer_TopLevelAssays/src/ARCtrl/ARCtrl.fs

**Comment by HLWeil** [comment]
Any ideas? @Freymaurer @kMutagene

File: src/ISA/ISA/ArcTypes/ArcTable.fs [hunk 1]
```diff
@@ -32,6 +32,14 @@ type ArcTable =
         Values = System.Collections.Generic.Dictionary<int*int,CompositeCell>()
     }
 
+    static member initWithHeaders(name,headers : ResizeArray<CompositeHeader>) =
```
Comment (Freymaurer) [inline src/ISA/ISA/ArcTypes/ArcTable.fs:35]: `createFromHeaders`

File: src/ISA/ISA/ArcTypes/ArcTable.fs [hunk 2]
```diff
@@ -156,7 +164,7 @@ type ArcTable =
         column.Cells |> Array.iteri (fun rowIndex v -> Unchecked.setCellAt(columnIndex,rowIndex,v) this.Values)
         Unchecked.fillMissingCells this.Headers this.Values
 
-    static member updatetColumn (columnIndex:int, header: CompositeHeader, ?cells: CompositeCell []) = 
+    static member updatedColumn (columnIndex:int, header: CompositeHeader, ?cells: CompositeCell []) =
```
Comment (Freymaurer) [inline src/ISA/ISA/ArcTypes/ArcTable.fs:167]: Please update to make difference to setColumn clear

File: src/ISA/ISA/ArcTypes/ArcTable.fs [hunk 3]
```diff
@@ -532,9 +540,64 @@ type ArcTable =
     static member fromProcesses name (ps : Process list) : ArcTable = 
         ps
         |> List.collect (fun p -> ProcessParsing.processToRows p)
-        |> fun rows -> ProcessParsing.alignByHeaders rows
+        |> fun rows -> ProcessParsing.alignByHeaders true rows
         |> fun (headers, rows) -> ArcTable.create(name,headers,rows)
 
+    /// This method is meant to update an ArcTable stored as a protocol in a study or investigation file with the information from an ArcTable actually stored as an annotation table
+    member this.UpdateReferenceByAnnotationTable(table:ArcTable) =
```
Comment (Freymaurer) [inline src/ISA/ISA/ArcTypes/ArcTable.fs:547]: Maybe move to a less discoverable location

File: src/ISA/ISA/ArcTypes/ArcTable.fs [hunk 4]
```diff
@@ -532,9 +540,64 @@ type ArcTable =
...
+            this.Headers
+            |> Seq.indexed
+            |> Seq.choose (fun (i,h) -> if h.isProtocolColumn then None else Some i)
+            |> Seq.toArray
+        this.RemoveColumns nonProtocolColumns
+        ArcTableAux.Unchecked.extendToRowCount table.RowCount this.Headers this.Values      
+        for c in table.Columns do
+            this.AddColumn(c.Header, cells = c.Cells,forceReplace = true)
+
+    static member SplitByColumnValues(columnIndex) =
```
Comment (Freymaurer) [inline src/ISA/ISA/ArcTypes/ArcTable.fs:558]: SplitByColumnValuesAt; and comment, that this divided table rows

File: src/ISA/ISA/ArcTypes/ArcTable.fs [hunk 5]
```diff
@@ -532,9 +540,64 @@ type ArcTable =
...
+            indexGroups
+            |> Array.mapi (fun i indexGroup ->
+                let headers  = table.Headers |> ResizeArray
+                let rows = 
+                    indexGroup
+                    |> Array.map (fun i -> table.GetRow(i))
+                ArcTable.createFromRows(table.Name,headers,rows)
+            )
+            
+    static member SplitByColumnValuesByHeader(header : CompositeHeader) =
```
Comment (Freymaurer) [inline src/ISA/ISA/ArcTypes/ArcTable.fs:571]: SplitByColumnValues, and add comment that this divides table by rows

File: src/ISA/ISA/ArcTypes/ArcTableAux.fs [hunk 6]
```diff
@@ -275,6 +278,16 @@ module Unchecked =
                 for missingColumn,missingRow in missingKeys do
                     setCellAt (missingColumn,missingRow,empty) values
 
+    /// Increases the table size to the given new row count and fills the new rows with the last value of the column
```
Comment (Freymaurer) [inline src/ISA/ISA/ArcTypes/ArcTableAux.fs:281]: Maybe combine with `fillMissingCells` and add a flag to use empty or last value

File: src/ISA/ISA/ArcTypes/ArcTables.fs [hunk 7]
```diff
@@ -297,8 +304,52 @@ type ArcTables(thisTables:ResizeArray<ArcTable>) =
...
             |> List.collect (fun p -> ProcessParsing.processToRows p)
-            |> fun rows -> ProcessParsing.alignByHeaders rows
+            |> fun rows -> ProcessParsing.alignByHeaders true rows
             |> fun (headers, rows) -> ArcTable.create(name,headers,rows)
         )
         |> ResizeArray
+        |> ArcTables
+
+    static member updateReferenceTablesBySheets (referenceTables : ArcTables,sheetTables : ArcTables,?keepUnusedRefTables : bool) : ArcTables =
+        let keepUnusedRefTables = Option.defaultValue false keepUnusedRefTables
```
Comment (Freymaurer) [inline src/ISA/ISA/ArcTypes/ArcTables.fs:314]: Can we move this somewhere less visible?

File: src/ISA/ISA/Update.fs [hunk 8]
```diff
@@ -1,5 +1,17 @@
 ﻿namespace ARCtrl.ISA.Aux
 
+module List =
```
Comment (Freymaurer) [inline src/ISA/ISA/Update.fs:3]: make this internal? So we never flood namespaces.

**Review by Freymaurer — COMMENTED** [review]
Please update naming and visibility issues ❤️

File: src/ISA/ISA/Update.fs (hunk 8 above)
Comment (HLWeil) [reply src/ISA/ISA/Update.fs:3]: Will first make a static member out of it. I think we need a place for more complex operations, that do not necessarily belong to ISA.Spreadsheet or ISA.Json.

File: src/ISA/ISA/ArcTypes/ArcTables.fs (hunk 7 above)
Comment (HLWeil) [reply src/ISA/ISA/ArcTypes/ArcTables.fs:314]: Will first make a static member out of it. We need a place for more complex operations that do not necessarily belong to a downstream repository.

File: src/ISA/ISA/ArcTypes/ArcTableAux.fs (hunk 6 above)
Comment (HLWeil) [reply src/ISA/ISA/ArcTypes/ArcTableAux.fs:281]: Now that I think about it, it doesn't really match. ExtendRows does not fill up empty cells in a previously sparse table, but extends the table to a new given rowCount.

File: src/ISA/ISA/ArcTypes/ArcTables.fs (hunk 7 above)
Comment (HLWeil) [reply src/ISA/ISA/ArcTypes/ArcTables.fs:314]: Whoops, this comment belongs to your comment above. Still my point remains.

Comment (Freymaurer) [reply src/ISA/ISA/ArcTypes/ArcTables.fs:314]: Maybe a module similiar to `module ARCtrl.ISA.IdentifierSetters` at the end of the arctypes folder with extensions?

**Review by Freymaurer — COMMENTED** [review]
👍 lgtm

**Pull request merged**